pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,184 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
+	}
+
+	if _, powerOffRequestExists := baremetalhost.Annotations[requestPowerOffAnnotation]; !powerOffRequestExists {
+		return &clustererror.RequeueAfterError{RequeueAfter: time.Second * 5}
+	}
+
+	delete(baremetalhost.Annotations, requestPowerOffAnnotation)
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to power-off request annotation from %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+// deleteMachineNode deletes the node that mapped to specified machine
+func (a *Actuator) deleteNode(ctx context.Context, node *corev1.Node) error {
+	err := a.client.Delete(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &clustererror.RequeueAfterError{}
+		}
+		log.Printf("Failed to delete node %s: %s", node.Name, err.Error())
+	}
+	return err
+}
+
+// getNodeByMachine returns the node object referenced by machine
+func (a *Actuator) getNodeByMachine(ctx context.Context, machine *machinev1.Machine) (*corev1.Node, error) {
+	if machine.Status.NodeRef == nil {
+		return nil, errors.NewNotFound(corev1.Resource("ObjectReference"), machine.Name)
+	}
+
+	node := &corev1.Node{}
+	key := client.ObjectKey{
+		Name:      machine.Status.NodeRef.Name,
+		Namespace: machine.Status.NodeRef.Namespace,
+	}
+
+	if err := a.client.Get(ctx, key, node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+/*
+remediateIfNeeded will try to remediate unhealthy machines (annotated by MHC) by power-cycle the host
+The full remediation flow is:
+1) Power off the host
+2) Add remediation-in-progress annotation to the unhealthy Machine
+3) Delete the node
+4) Power on the host
+5) Wait for the node the come up
+6) Remove remediation-in-progress annotation and the annotation added by MAO to signal the machine is unhealthy
+ */
+func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1.Machine, baremetalhost *bmh.BareMetalHost) error {
+	if len(machine.Annotations) == 0 {
+		return nil
+	}
+
+	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
+		return nil
+	}
+
+	if _, remediationInProgress := machine.Annotations[remediationInProgressAnnotation]; !remediationInProgress {
+		if !hasPowerOffRequestAnnotation(baremetalhost) {
+			log.Printf("Found an unhealthy machine, requesting power off. Machine name: %s", machine.Name)
+			return a.requestPowerOff(ctx, baremetalhost)
+		}
+
+		//hold remediation until the power off request is fulfilled
+		if baremetalhost.Status.PoweredOn {
+			return nil
+		}
+
+		//we need this annotation to differentiate between unhealthy machine that
+		//needs remediation, and an unhealthy machine that just got remediated
+		return a.addRemediationInProgressAnnotation(ctx, machine)
+	}
+
+	node, err := a.getNodeByMachine(ctx, machine)
+
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Printf("Failed to get Node from Machine %s: %s", machine.Name, err.Error())
+			return err
+		}
+	}
+
```

Comment:
The logic below this point is hard for me to follow. It seems to be tracking 2 conditions (whether the node is nil and whether the host is powered on), but it's not doing it in a way that makes clear what transitions are actually of interest, or why. What does it mean for the system to be in each of the 4 states?

The first conditional block only applies to existing nodes and the next one applies to nil nodes but both look for a host that is powered off. The test for node != nil on line 770 would seem redundant, except if we end up with a non-nil node that is powered on that logic will be triggered and that doesn't seem correct.

Maybe having some inline comments would help.

Reply:
I think that if you follow the docs I've added before the function declaration it could make more sense:
1) Power off the host
2) Add remediation-in-progress annotation to the unhealthy Machine
3) Delete the node
4) Power on the host
5) Wait for the node the come up
6) Remove remediation-in-progress annotation and the annotation added by MAO to signal the machine is unhealthy

But I can understand the confusion, and I'll try to add some more comments on each step

> The test for node != nil on line 770 would seem redundant, except if we end up with a non-nil node that is powered on that logic will be triggered and that doesn't seem correct.

Why this doesn't seem correct?
We want to end up with powered-on host with available node, this means we finished remediation. Hence we delete the unhealthy annotation added by MHC, and removing the `remediation-in-progress` annotation